refactor(github): redo GitHub workflow definitions

zregvart opened the pull request:
Seems that we have accumulated a lot of complexity, unneeded duplication
and straight up unnecessary configuration in the GitHub workflow
definitions. This simplifies, optimizes and reduces the complexity of
those.

We have a single workflow `checks.yml` that applies to both pull
requests and to release branches (`N.M.x`).

Work done is minimized by checking what is changed, and build/checks are
only run for the subset of changes. For any change in the `docs/**` we
don't run any check; backend, operator and ui-react are built only if
changes have been made within their specific directories; images are
built for any change in the backend, operator or ui-react to keep the
image set consistent; and lastly integration tests are run for any
change within the backend.

fix(tools): don't build operator images by default (d925acd)

The convention within the `syndesis` tool is that if no argument is
given for an image to be built the build of the image should not be
performed. This extends this convention to the operator build as well.

zregvart committed:
refactor(github): redo GitHub workflow definitions

Seems that we have accumulated a lot of complexity, unneeded duplication
and straight up unnecessary configuration in the GitHub workflow
definitions. This simplifies, optimizes and reduces the complexity of
those.

We have a single workflow `checks.yml` that applies to both pull
requests and to release branches (`N.M.x`).

Work done is minimized by checking what is changed, and build/checks are
only run for the subset of changes. For any change in the `docs/**` we
don't run any check; backend, operator and ui-react are built only if
changes have been made within their specific directories; images are
built for any change in the backend, operator or ui-react to keep the
image set consistent; and lastly integration tests are run for any
change within the backend.

zregvart committed:
fix(tools): don't build operator images by default

The convention within the `syndesis` tool is that if no argument is
given for an image to be built the build of the image should not be
performed. This extends this convention to the operator build as well.

zregvart commented:
I've tested this on my fork with:
 - [docs change](https://github.com/zregvart/syndesis/pull/9)
 - [ui change](https://github.com/zregvart/syndesis/actions/runs/767049502)
 - [backend change](https://github.com/zregvart/syndesis/actions/runs/767059654)
 - [operator change](https://github.com/zregvart/syndesis/actions/runs/767071222)

We can wait and see how those checks fare.

Delawen commented on .github/workflows/checks.yml:
```diff
@@ -0,0 +1,191 @@
+        run: tools/bin/syndesis build --batch-mode --backend
+      - name: Publish Test Reports
+        if: always()
+        uses: scacap/action-surefire-report@v1
```
Good idea!

Delawen commented on .github/workflows/checks.yml:
```diff
@@ -0,0 +1,191 @@
+      ui: ${{ steps.filter.outputs.ui }}
+    steps:
+      - uses: actions/checkout@v2
+        if: needs.setup.outputs.isDefaultBranch == 'true'
```
So if it is not the default branch... it does not run anything at all? Because the changes.outpus.* will be false if this is not run, right?

zregvart replied:
From what I understood in the https://github.com/dorny/paths-filter action needs a git repository only for repository branches, not for pull requests. So this skips the https://github.com/actions/checkout for non-default branches (pull requests and every other branch). This could be expanded to include other branches from the repository. This was a first pass at this, and it'll work for `1.12.x`.

zregvart committed:
fix(github): add missing `IMAGES_DIR` env variable

zregvart commented:
I'm trying to fix the issue that running `syndesis build -m ui-react` is not possible without previously running `syndesis build -m common` (as seen in the https://github.com/zregvart/syndesis/runs/2390379579?check_suite_focus=true). For this I need to de-activate the `non-staging` profile based on existance of `stagingDescription` **AND** `build-selection` properties. This is not possible in Maven out of the box and I'm looking into creating an extension to help with that. There are other extensions available, I've tried 3 or 4 of them, none of them work.

zregvart committed:
chore(build): tweak the non-release profile

The `non-release` profile, previously known as `non-staging` profile is
a profile that's meant to exclude certain modules form building in a
release build. With changes to the syndesis command and the profile
refactoring the profile was built also for modules that do not need it,
such as `ui-react`.
This instigates an extension that allows us to select profiles based on
an expression, which in this case is that the `release` profile is not
active and that the `build-selection`, i.e. building specific modules,
is not done.

zregvart committed:
fix(github): make sure `IMAGES_DIR` is set

zregvart commented:
All workflows succeeded on my fork:
 - [1.12.x](https://github.com/zregvart/syndesis/actions/runs/775214831)
 - [UI change](https://github.com/zregvart/syndesis/actions/runs/775393965)
 - [backend change](https://github.com/zregvart/syndesis/actions/runs/775397678)
 - [operator change](https://github.com/zregvart/syndesis/actions/runs/775398981)

I think we can merge this now. WDYT @phantomjinx @Delawen cc @avano
